### PR TITLE
Dead code removal and bring code coverage up

### LINKED ISSUE #152: Prune dead code
There appears to be code that isn't using anymore and is only covered by tests:

* `type_equation_solver.py`
* `coercion.py`

I think we should remove it if it isn't being used by any Continuum projects that depend on datashape.

### PR DESCRIPTION
**Pull request by cpcloud**
closes #152 

- [x] remove `overload_resolver.py`, `coercion.py`
- [x] unify `promotion.py` and `promote.py`
- [x] use `TypeError` in favor of `NotNumpyCompatible`
- [x] add a lot of test coverage for small things like repr and larger-ish things like testing that we raise errors correctly
- [x] Remove the DataShapeException hierarchy in favor of using the exceptions in the stdlib
- [x] Remove a large amount of dead code

### CODE REVIEW

**Comment by cpcloud** [comment]
@llllllllll can you take a look at this? sorry for the huge changeset, i wanted to rip the bandaid off all at once and just get it over with.

i've tested locally against blaze and odo and things are working FWIW

File: datashape/coretypes.py [hunk 1]
```diff
@@ -526,18 +455,10 @@ def __init__(self, *parameters, **kwds):
...
-            self.name = name
-            self.__metaclass__._registry[name] = self
-        else:
-            self.name = None
-
-        ###
-        # TODO: Why are low-level concepts like strides and alignment on
-        # TODO: the datashape?
-        ###
+        if self.name is not None:
```
Comment (llllllllll) [inline datashape/coretypes.py]: this will change behaviour if name is `''`, is this okay?

File: datashape/discovery.py [hunk 2]
```diff
@@ -160,13 +172,10 @@ def discover(s):
...
-        if not d.time():
-            return date_
-        if not d.date():
-            return time_
-        return datetime_
-    except:
+    except ValueError:
         pass
+    else:
+        return date_ if not d.time() else datetime_
```
Comment (llllllllll) [inline datashape/discovery.py]: this is missing `if not d.date(): return time_`

File: datashape/predicates.py [hunk 3]
```diff
@@ -104,7 +109,6 @@ def _dimensions(ds):
         return 1 + _dimensions(ds.subshape[0])
     if isscalar(ds):
         return 0
-    raise NotImplementedError('Can not compute dimensions for %s' % ds)
```
Comment (llllllllll) [inline datashape/predicates.py]: should this just return `None` if none of these checks catch? If `None` is valid, then we should have an explicit return here. If at least one of branches is expected, then we should replace this with an `AssertionError` so that it is easier to debug in the future.

File: datashape/discovery.py (hunk 2 above)
Comment (cpcloud) [inline datashape/discovery.py]: how does that ever happen? how do you construct a `datetime.date` object that is falsey?

Comment (llllllllll) [inline datashape/discovery.py]: ah, I see. nvm then

File: datashape/tests/test_parser.py [hunk 4]
```diff
@@ -5,13 +5,21 @@
...
 from datashape import coretypes as ct
 from datashape import DataShapeSyntaxError
 
+
+@pytest.fixture
+def sym():
+    return datashape.TypeSymbolTable()
+
+
 class TestDataShapeParseBasicDType(unittest.TestCase):
```
Comment (llllllllll) [inline datashape/tests/test_parser.py]: is there a reason we are mixing `py.test` with `unittest` style test cases in this file?

File: datashape/tests/test_util.py [hunk 5]
```diff
@@ -1,91 +1,72 @@
...
-        self.assertEqual(datashape.cat_dshapes(dslist),
-                        dslist[0])
-        # two dshapes
-        dslist = [dshape('3 * 10 * int32'),
-                        dshape('7 * 10 * int32')]
-        self.assertEqual(datashape.cat_dshapes(dslist),
-                        dshape('10 * 10 * int32'))
+def test_cat_dshapes():
+    # concatenating 1 dshape is a no-op
+    dslist = [dshape('3 * 10 * int32')]
```
Comment (llllllllll) [inline datashape/tests/test_util.py]: I didn't realize this was in datashape. I should refactor `Concat` to use this.

File: datashape/tests/test_parser.py (hunk 4 above)
Comment (cpcloud) [inline datashape/tests/test_parser.py]: because i didn't want to rewrite those yet ... that'll be in another PR

File: datashape/coretypes.py (hunk 1 above)
Comment (cpcloud) [inline datashape/coretypes.py]: good point, i'll change that back to `if self.name`

File: datashape/tests/test_parser.py (hunk 4 above)
Comment (llllllllll) [inline datashape/tests/test_parser.py]: kk, just wanted to make sure it wasn't for another reason

File: datashape/typesets.py [hunk 6]
```diff
@@ -90,8 +89,9 @@ def __init__(self):
         self.lookup = self.registry.get
 
     def register_typeset(self, name, typeset):
-        if name in typeset:
-            raise DataShapeTypeError("TypeSet %s already defined" % name)
+        if name in self.registry:
```
Comment (llllllllll) [inline datashape/typesets.py]: this looks more correct; however, I just want to make sure that the behaviour change is okay here.

File: datashape/coretypes.py [hunk 7]
```diff
@@ -1217,31 +1109,20 @@ def to_numpy(ds):
...
-        dtype = msr.to_numpy_dtype()
-    except AttributeError:
-        raise NotNumpyCompatible('DataShape measure %s is not NumPy-compatible'
-                                 % msr)
-
-    if type(dtype) != np.dtype:
-        raise NotNumpyCompatible('Internal Error: Failed to produce NumPy '
-                                 'dtype')
-    return shape, dtype
+    return shape, msr.to_numpy_dtype()
```
Comment (llllllllll) [inline datashape/coretypes.py]: should we continue to catch the potential `AttributeError` here to reraise as a `TypeError`?
Maybe something like
```python
try:
    to_numpy_dtype = msr.to_numpy_dtype
except AttributeError:
    raise TypeError('DataShape measure %s is not NumPy-compatible' % msr)

return shape, to_numpy_dtype()
```

Comment (cpcloud) [inline datashape/coretypes.py]: okay, i think i'd rather just have a default implementation of this and have it raise rather than catching an `AttributeError`

Comment (llllllllll) [inline datashape/coretypes.py]: sounds good

File: datashape/typesets.py (hunk 6 above)
Comment (cpcloud) [inline datashape/typesets.py]: `name in typeset` is a bug. we're registering the typeset and typesets contain other dshapes, so it doesn't make sense to look for a string contained in the typeset we're registering.

File: datashape/predicates.py (hunk 3 above)
Comment (cpcloud) [inline datashape/predicates.py]: i forgot to add a test that exercises this code and i saw that it wasn't being hit. it's definitely possible to get here and we don't necessarily expect one of the branches to return. a function prototype is an example of something that doesn't have dimensions

File: datashape/coretypes.py (hunk 1 above)
Comment (cpcloud) [inline datashape/coretypes.py]: i thought about removing this feature entirely, but it's used quite heavily in `CType`

File: datashape/typesets.py (hunk 6 above)
Comment (llllllllll) [inline datashape/typesets.py]: kk

File: datashape/predicates.py (hunk 3 above)
Comment (cpcloud) [inline datashape/predicates.py]: done

**Comment by cpcloud** [comment]
@llllllllll any more comments?

**Comment by cpcloud** [comment]
merging if none

**Comment by llllllllll** [comment]
no more, merge away
